storage: never elide resume key in MVCCExportToSST

stevendanna opened the pull request:
MVCCExportToSST checks if no data has been written to the SST and avoids calling Finish() on the writer in that case. However, it was previously also returning an empty resume key in that case, regardless of the resume key's previous value.

Returning an empty resume key was OK since it was assumed that if the SST had no data, then there couldn't possibly be anything to resume.

However, when this method was extended to support cooperative resource management, it now has two cases in which the main iteration loop of the export may return a resume key despite not having written any data.

First, the Elastic CPU work handler is called unconditionally. If the CPU work handler was already over the limit on the first call, then we return immediately and with no data and a resume key set.

Second, the time-based resource limiter is called after the first iteration of the loop. While typically allowing one iteration will result in data being written to the SST before the exhaustion condition can be hit; if there is a long string of deletion tombstones that are not being added to our response, then it is possible that we've completed iterations of the loop without any data being written.

As a result of dropping this resume key, users of ExportRequest such as backup would then assume that their work was done. As a result, a backup may not contain all of the expected data.

Elastic CPU limiting and the time-based resource limiter are disabled by default in previous releases.

Elastic CPU limiting is enabled by default on master.

Epic: None

Release note (bug fix): Fix a bug that would result in incomplete backups when non-default, non-public resource limiting settings (kv.bulk_sst.max_request_time or
admission.elastic_cpu.enabled) were enabled.

stevendanna committed:
storage: never elide resume key in MVCCExportToSST

MVCCExportToSST checks if no data has been written to the SST and
avoids calling Finish() on the writer in that case. However, it was
previously also returning an empty resume key in that case, regardless
of the resume key's previous value.

Returning an empty resume key was OK since it was assumed that if the
SST had no data, then there couldn't possibly be anything to resume.

However, when this method was extended to support cooperative resource
management, it now has two cases in which the main iteration loop of
the export may return a resume key despite not having written any
data.

First, the Elastic CPU work handler is called unconditionally. If the
CPU work handler was already over the limit on the first call, then we
return immediately and with no data and a resume key set.

Second, the time-based resource limiter is called after the first
iteration of the loop. While typically allowing one iteration will
result in data being written to the SST before the exhaustion
condition can be hit; if there is a long string of deletion tombstones
that are not being added to our response, then it is possible that
we've completed iterations of the loop without any data being written.

As a result of dropping this resume key, users of ExportRequest such
as backup would then assume that their work was done. As a result, a
backup may not contain all of the expected data.

Elastic CPU limiting and the time-based resource limiter are disabled
by default in previous releases.

Elastic CPU limiting is enabled by default on master.

Release note (bug fix): Fix a bug that would result in incomplete
backups when non-default, non-public resource limiting
settings (kv.bulk_sst.max_request_time or
admission.elastic_cpu.enabled) were enabled.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/92825)
<!-- Reviewable:end -->

stevendanna commented:
As @adityamaru pointed out we also have to update cmd_export.go

stevendanna committed:
kvserver: respect resume key even when DataSize is 0

Release note: None

irfansharif commented on pkg/storage/mvcc_test.go:
```diff
@@ -5961,6 +5963,149 @@ func TestMVCCExportToSSTResourceLimits(t *testing.T) {
 
+// TestMVCCExportToSSTExhaustedAtStart is a regression test for a bug
+// in which mis-handling of resume spans would cause MVCCExportToSST
+// would return an empty resume key in cases where out various
```
Something's funky with this sentence around "would return", s/out/our.

irfansharif commented on pkg/storage/mvcc.go:
```diff
@@ -5866,9 +5866,9 @@ func MVCCExportToSST(
-		// nil slice – the export code will discard it anyway (based on 0 DataSize).
-		return roachpb.BulkOpSummary{}, MVCCKey{}, nil
+		// If no records were added to the sstable, skip
+		// completing it and return an empty summary.
```
Perhaps add something about why we're propagating the resume key (what you've explained in the commit message -- we might still want resumption).

irfansharif reviewed (approved):
Amazing catch, how'd you find it? LGTM. I should've double checked when cargo culting from the time-based resource limiter. I'll update some internal guidance around not enabling this machinery for 22.2 clusters until patch is in, probably 22.2.1 at this point (?).

sumeerbhola reviewed (changes requested):
Nice catch!

> Reviewed 1 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/92825#-:-NIIE8Qe7B_5_4xCThpZ:b-73m6vw)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif and @stevendanna)

___
*[`pkg/storage/mvcc.go` line 6009 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/92825#-NIIBBwu0k2ey4N5Dz_G:-NIIBBwu0k2ey4N5Dz_H:be29lyq) ([raw file](https://github.com/cockroachdb/cockroach/blob/be9512b1cef37899dd5e9c0869593c5efe347d3f/pkg/storage/mvcc.go#L6009)):*
> ```Go
> 		unsafeKey := iter.UnsafeKey()
> 
> 		isNewKey := !opts.ExportAllRevisions || !unsafeKey.Key.Equal(curKey)
> ```

The code in this loop is overcomplicated. I am not sure what if anything can be simplified, but documenting some invariants would make it easier to be sure of correctness (and some RaceEnabled invariant checking).
I started writing the following comment here
```
		// !opts.ExportAllRevisions => isNewKey since we call iter.NextKey below
		// that will skip older versions.
		//
		// When opts.ExportAllRevisions, we need to compare with curKey to compute
		// isNewKey. For the first loop iteration, curKey is empty, which implies
		// isNewKey is true (at least from the perspective of this call to
		// mvccExportToWriter, ignoring the sequence of such calls that were
		// resumed). For subsequent loop iterations the correctness of this
		// comparison relies on saving curKey. Note that there is an additional
		// conjunct (trackKeyBoundary) that gates the saving of the key which is
		// computed as paginated || opts.ResourceLimiter != nil. If that conjunct
		// is not true, this key comparison is bogus, so the comparison result must
		// only be used in contexts where that conjunct is true.
```

And then realized there may be a bug. The `if overLimit, _ := elasticCPUHandle.OverLimit(); overLimit && isNewKey {
` block can execute even if `paginated` is false and `opts.ResourceLimiter == nil`. In which case `isNewKey` has been computed using a `curKey` that has never been updated, so will always be true. Which means we could return with a subset of versions.

Slightly unrelated, it's not clear to me why we are not using the `first` logic that is there for the `ResourceLimiter` checking, also in the elastic cpu checking. I realize walltime is more likely to be exceeded, due to the goroutine not being scheduled, and the `ElasticCPUWorkHandle` is counting actual running time. But there has been work done before the first iteration, and since the `ElasticCPUWorkHandle` was created -- latching, locking, and most notably the `SeekGE` before the loop. There is a non-zero risk that we will return without making any progress.
 


<!-- Sent from Reviewable.io -->

stevendanna commented:
TFTR!

> how'd you find it?

I've been writing a "backup nemesis" test for incremental backups and then running it under stress. During that I was seeing failure related to this, but it took me a while to rule out that my new test wasn't just incorrect.  Now that we understand this, I think it may also explain a couple of other unexplained failures we've seen when stressing tests.

stevendanna committed:
storage: fix more resume key handling

Previously, the elastic CPU work handler could result in a returned
SST with only a portion of the key's revisions, regardless of the
value of StopMidKey.

Now, we correctly track the previous key when a non-nil elastic CPU
work handler is provided.

Further, we would previously return an empty export request in the
case where the first key we tried to add exceeded the max size. Now,
return an error in that case.

Release note: None

stevendanna commented:
> The code in this loop is overcomplicated. 

I agree. To fix the bug you mentioned below, I needed to track the key boundary whenever elasticCPUHandle is over the limit.  As a result, I think we are almost always tracking the key boundary. So, perhaps one way to simplify this would be to live with that fact and thus always track the key boundary.

In a follow-up I'm also going to remove the legacy resource limiter.

Some larger changes that might make this simpler would require some thinking around backwards compatibility. But it seems like we might want to make StopMidKey mandatory of anyone requesting paginated results.

> And then realized there may be a bug. The if overLimit, _ := elasticCPUHandle.OverLimit(); overLimit && isNewKey { block can execute even if paginated is false and opts.ResourceLimiter == nil. In which case isNewKey has been computed using a curKey that has never been updated, so will always be true. Which means we could return with a subset of versions.

Nice catch. I believe I've fixed a couple of other bugs:

- we would return mid key even if StopMidKey was false,
- we would return an empty SST with a resume key that was destined to fail again if the first key exceeded the max size.

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 3 files at r1, 3 of 5 files at r4, 1 of 1 files at r5, 4 of 4 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/92825#-:-NIu3vpV8dS3wyDl565k:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @stevendanna)

___
*[`pkg/storage/mvcc.go` line 6056 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/92825#-NIu3X9Q3b_it-mKT9lm:-NIu3X9Q3b_it-mKT9ln:b-f7d416) ([raw file](https://github.com/cockroachdb/cockroach/blob/c857ec4bdd5337f2a3128aff06de69b6a804f6e3/pkg/storage/mvcc.go#L6056)):*
> ```Go
> 
> 		isNewKey := !opts.ExportAllRevisions || !unsafeKey.Key.Equal(curKey)
> 		if trackKeyBoundary && opts.ExportAllRevisions && isNewKey {
> ```

I think we could still use a comment like the one I wrote stating some code invariants.
___
*[`pkg/util/admission/elastic_cpu_work_handle.go` line 78 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/92825#-NIu1OBA8od5uSDotxRO:-NIu1OBA8od5uSDotxRP:ba2y3nq) ([raw file](https://github.com/cockroachdb/cockroach/blob/505f290a6c7a53004a9eff9fb51c32a198679261/pkg/util/admission/elastic_cpu_work_handle.go#L78)):*
> ```Go
> 	// TODO(ssd): It would be nice to get this in some zero-cost
> 	// way. We could make ElasticCPUWorkHandle an interface and
> 	// then allow for testing implementations. Not sure if the
> ```

This is very performance sensitive so I don't think we would want an interface.
Can the test code call a test-only method to set itersUntilCheck to 0. Then we could move this conditional into `overLimitInner` where I suspect we can more easily afford the cost.

@irfansharif may have other ideas.


<!-- Sent from Reviewable.io -->

stevendanna committed:
storage: move elastic CPU test hook out of fast path

Epic: none
Release note: None

stevendanna committed:
storage: clear ts before returning resume key when on new key

This matches the behaviour with our other early exits (except in the
case of range keys). We use the timestamp on this key to determine
whether we've already started splitting mid key.

This seems unnecessary to me, but I need to think about it some more.

Epic: None
Release note: None

stevendanna commented:
bors r=sumeerbhola,irfansharif
